internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+	if err != nil {
+		return time.Now(), err
+	}
+
+	var now time.Time
+	err = mds.db.QueryRowContext(datastore.SeparateContextWithTracing(ctx), nowSQL, nowArgs...).Scan(&now)
+	if err != nil {
+		return time.Now(), err
+	}
+
+	// RelationTupleTransaction is not timezone aware
+	// Explicitly use UTC before using as a query arg
+	now = now.UTC()
+
+	return now, nil
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+// - this implementation does not have metrics yet
+// - an additional useful logging message is added
+// - context is removed from logger because zerolog expects logger in the context
+func (mds *Datastore) collectGarbage() error {
+	ctx, cancel := context.WithTimeout(context.Background(), mds.gcMaxOperationTime)
+	defer cancel()
+
+	// Ensure the database is ready.
+	ready, err := mds.IsReady(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !ready {
+		log.Warn().Msg("cannot perform mysql garbage collection: mysql driver is not yet ready")
+		return nil
+	}
+
+	now, err := mds.getNow(ctx)
+	if err != nil {
+		return err
+	}
+
+	before := now.Add(mds.gcWindowInverted)
+	log.Debug().Time("before", before).Msg("running mysql garbage collection")
+	relCount, transCount, err := mds.collectGarbageBefore(ctx, before)
+	log.Debug().Int64("relationshipsDeleted", relCount).Int64("transactionsDeleted", transCount).Msg("garbage collection completed for mysql")
+	return err
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+// - main difference is how the PSQL driver handles null values
+func (mds *Datastore) collectGarbageBefore(ctx context.Context, before time.Time) (int64, int64, error) {
+	// Find the highest transaction ID before the GC window.
+	query, args, err := mds.GetRevision.Where(sq.Lt{colTimestamp: before}).ToSql()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	var value sql.NullInt64
+	err = mds.db.QueryRowContext(
+		datastore.SeparateContextWithTracing(ctx), query, args...,
+	).Scan(&value)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if !value.Valid {
+		log.Debug().Time("before", before).Msg("no stale transactions found in the datastore")
+		return 0, 0, nil
+	}
+	highest := uint64(value.Int64)
+
+	log.Trace().Uint64("highestTransactionId", highest).Msg("retrieved transaction ID for GC")
+
+	return mds.collectGarbageForTransaction(ctx, highest)
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+// - implementation misses metrics
+func (mds *Datastore) collectGarbageForTransaction(ctx context.Context, highest uint64) (int64, int64, error) {
+	// Delete any relationship rows with deleted_transaction <= the transaction ID.
+	relCount, err := mds.batchDelete(ctx, mds.driver.RelationTuple(), sq.LtOrEq{colDeletedTxn: highest})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	log.Trace().Uint64("highestTransactionId", highest).Int64("relationshipsDeleted", relCount).Msg("deleted stale relationships")
+	// Delete all transaction rows with ID < the transaction ID. We don't delete the transaction
```

Comment:
newline before this line

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/a353ef91a70f1424251c5c26ebdf2114ed53c622